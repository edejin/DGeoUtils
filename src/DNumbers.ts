import {DEGREE_TO_PI, DOUBLE_PI_IN_DEGREE, PI_IN_DEGREE, PI_TO_DEGREE} from './DPoint';

const delta = 0.001;

export class DNumbers {
  static like(v: number, s: number, d: number = delta): boolean {
    return Math.abs(v - s) < d;
  }

  static likeZero(v: number): boolean {
    return DNumbers.like(v, 0);
  }

  static like2PI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), DOUBLE_PI_IN_DEGREE);
  }

  static likePI(v: number): boolean {
    return DNumbers.like(DNumbers.rad2Deg(v), PI_IN_DEGREE);
  }

  static rad2Deg(v: number): number {
    return v * DEGREE_TO_PI;
  }

  static deg2Rad(v: number): number {
    return v * PI_TO_DEGREE;
  }
}
