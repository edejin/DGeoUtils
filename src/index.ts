export * from './DCircle';
export * from './DLine';
export * from './DNumbers';
export * from './DPoint';
export * from './DPolygon';
export * from './FastSearch';
export * from './TraceMatrix';
