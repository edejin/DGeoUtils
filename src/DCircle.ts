import {DPoint, EARTH_RADIUS_IN_METERS} from './DPoint';
import {DPolygon} from './DPolygon';
import {DNumbers} from './DNumbers';
import {checkFunction} from './utils';

// eslint-disable-next-line padded-blocks
export class DCircle {

  /**
   * @param [center=(0,0)]
   * @param [r=0]
   */
  // eslint-disable-next-line no-useless-constructor,no-empty-function
  constructor(public center: DPoint = DPoint.zero(), public r: number = 0) {}

  toString(): string {
    return `(${this.center.toString()}, ${this.r})`;
  }

  getValue(): {center: DPoint, r: number} {
    return {center: this.center, r: this.r};
  }

  clone(): DCircle {
    return new DCircle(this.center, this.r);
  }

  /**
   * Find intersection points with other circle.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPoints.png)
   * @param c
   */
  findPoints(c: DCircle): DPoint[] | number {
    if (this.equal(c)) {
      return Infinity;
    }
    const {
      center: {x: x0, y: y0},
      r: r0
    } = this;
    const {
      center: {x: x1, y: y1},
      r: r1
    } = c;

    const r02 = r0 * r0;

    const d = this.center.distance(c.center);
    const a = (r02 - r1 * r1 + d * d) / (2 * d);
    const h = Math.sqrt(r02 - a * a);

    const ad = a / d;
    const dy = y1 - y0;
    const dx = x1 - x0;
    const hd = h / d;

    const x2 = x0 + ad * (x1 - x0);
    const y2 = y0 + ad * (y1 - y0);

    const x31 = x2 + hd * dy;
    const y31 = y2 - hd * dx;
    const x32 = x2 - hd * dy;
    const y32 = y2 + hd * dx;

    const res = [];
    if (!isNaN(x31) && !isNaN(y31)) {
      res.push(new DPoint(x31, y31));
    }
    if (!isNaN(x32) && !isNaN(y32) && !(x31 === x32 && y31 === y32)) {
      res.push(new DPoint(x32, y32));
    }
    return res;
  }

  equal({center, r}: DCircle): boolean {
    return this.center.equal(center) && this.r === r;
  }

  /**
   * Transform circle to polygon
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInside.png)
   * @param [pointCount=64]
   */
  findPolygonInside(pointCount: number = 64): DPolygon {
    const preAngle = 2 * Math.PI / pointCount;
    const points: DPoint[] = [];
    for (let i = 0; i < pointCount; i++) {
      const angle = preAngle * i;
      const x = this.r * Math.cos(angle) + this.center.x;
      const y = this.r * Math.sin(angle) + this.center.y;
      points.push(new DPoint(x, y));
    }
    return new DPolygon([...points, points[0]]);
  }

  /**
   * Transform circle to polygon on sphere. It would be different for different latitude.
   *
   * @remarks
   * Center should be Lng/Lat.
   *
   * @remarks
   * Radius should be in meters.
   *
   * ![Example](https://edejin.github.io/DGeoUtils/media/examples/findPolygonInsideOnSphere.png)
   * @param [pointCount=64]
   */
  findPolygonInsideOnSphere(pointCount: number = 64): DPolygon {
    checkFunction('findPolygonInsideOnSphere')
      .checkArgument('center')
      .shouldBeDegree(this.center);
    const res = new DPolygon();
    for (let i = 0; i < pointCount; i++) {
      res.push(this.sphereOffset(2 * Math.PI * i / pointCount));
    }
    return res.close();
  }

  private sphereOffset(bearing: number, earthRadius = EARTH_RADIUS_IN_METERS): DPoint {
    const lat1 = DNumbers.deg2Rad(this.center.y);
    const lon1 = DNumbers.deg2Rad(this.center.x);
    const dByR = this.r / earthRadius;
    const lat = Math.asin(Math.sin(lat1) * Math.cos(dByR) +
      Math.cos(lat1) * Math.sin(dByR) * Math.cos(bearing));
    const lon =
      lon1 +
      Math.atan2(
        Math.sin(bearing) * Math.sin(dByR) * Math.cos(lat1),
        Math.cos(dByR) - Math.sin(lat1) * Math.sin(lat)
      );
    return new DPoint(DNumbers.rad2Deg(lon), DNumbers.rad2Deg(lat));
  }
}
