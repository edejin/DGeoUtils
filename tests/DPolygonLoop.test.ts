import {DPoint, DPolygon} from '../src';

// eslint-disable-next-line max-lines-per-function
describe('DPolygonLoop', () => {
  describe('basic', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p
        .loop()
        .move(1)
        .scale(2)
        .run()
        .toString())
        .toBe('(4 6, 8 10)');
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p
        .loop()
        .scale(2)
        .move(1)
        .run()
        .toString())
        .toBe('(3 5, 7 9)');
    });
    test('3', () => {
      expect(DPolygon.parseFromWKT('POLYGON ((30.1 10.2, 40.3 40.4, 20.5 40.6, 10.7 20.8, 30.1 10.2),' +
        ' (25.2 18.3, 30.4 31.5, 21.6 32.7, 18.8 22.9, 25.1 18.2))')
        .loop()
        .move(1)
        .scale(2)
        .run()
        .toWKT()).toBe('POLYGON ((62.2 22.4, 82.6 82.8, 43 83.2, 23.4 43.6, 62.2 22.4),' +
        ' (52.4 38.6, 62.8 65, 45.2 67.4, 39.6 47.8, 52.2 38.4))');
    });
    test('4', () => {
      expect(DPolygon.parseFromWKT('POLYGON ((30.1 10.2, 40.3 40.4, 20.5 40.6, 10.7 20.8, 30.1 10.2),' +
        ' (25.2 18.3, 30.4 31.5, 21.6 32.7, 18.8 22.9, 25.1 18.2))')
        .loop()
        .scale(2)
        .move(1)
        .run()
        .toWKT()).toBe('POLYGON ((61.2 21.4, 81.6 81.8, 42 82.2, 22.4 42.6, 61.2 21.4),' +
        ' (51.4 37.6, 61.8 64, 44.2 66.4, 38.6 46.8, 51.2 37.4))');
    });
  });
  describe('move', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p
        .loop()
        .move(1)
        .run()
        .toString())
        .toBe('(2 3, 4 5)');
    });
  });
  describe('scale', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p
        .loop()
        .scale(2)
        .run()
        .toString())
        .toBe('(2 4, 6 8)');
    });
  });
  describe('getTileFromCoords', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0),
        new DPoint(33.6, 44.5),
        new DPoint(54, 24)
      ]);
      expect(p.loop()
        .getTileFromCoords(13)
        .run()
        .toString())
        .toBe('(4096 4096, 4860 2962, 5324 3533)');
    });
  });
  describe('getCoordsFromTile', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .getCoordsFromTile()
        .run()
        .toString())
        .toBe('(0 0, 33.57421875 44.5278427984555, 53.96484375 24.00632619875113)');
    });
  });
  describe('transform', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0),
        new DPoint(33.6, 44.5),
        new DPoint(54, 24)
      ]);
      expect(p.loop()
        .metersToDegree()
        .run()
        .toString())
        .toBe('(0 0, 0.0003018339355061748 0.00039975030148298174,' +
          ' 0.00048509025349206667 0.00021559566822304532)');
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(0, 0),
        new DPoint(33.6, 44.5),
        new DPoint(54, 24)
      ]);
      expect(p.loop()
        .degreeToMeters()
        .run()
        .toString())
        .toBe('(0 -7.081154550627917e-10, 3740334.8901333325 5543147.203090187,' +
          ' 6011252.501999999 2753408.1089817025)');
    });
  });
  describe('height', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .height(14)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [4096, 4096, 14],
          [4860, 2962, 14],
          [5324, 3533, 14]
        ]);
    });
  });
  describe('setX', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setX(14)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [14, 4096, 13],
          [14, 2962, 13],
          [14, 3533, 13]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setX(({
          x,
          y
        }: DPoint) => x + y)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [8192, 4096, 13],
          [7822, 2962, 13],
          [8857, 3533, 13]
        ]);
    });
  });
  describe('setY', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setY(14)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [4096, 14, 13],
          [4860, 14, 13],
          [5324, 14, 13]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setY(({
          x,
          y
        }: DPoint) => x + y)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [4096, 8192, 13],
          [4860, 7822, 13],
          [5324, 8857, 13]
        ]);
    });
  });
  describe('setZ', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setZ(14)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [4096, 4096, 14],
          [4860, 2962, 14],
          [5324, 3533, 14]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(4096, 4096, 13),
        new DPoint(4860, 2962, 13),
        new DPoint(5324, 3533, 13)
      ]);
      expect(p.loop()
        .setZ(({
          x,
          y
        }: DPoint) => x + y)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [4096, 4096, 8192],
          [4860, 2962, 7822],
          [5324, 3533, 8857]
        ]);
    });
  });
  describe('rotate', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(30, 30),
        new DPoint(60, 30),
        new DPoint(60, 60),
        new DPoint(30, 60)
      ]);
      expect(p
        .loop()
        .rotate(Math.PI / 6)
        .toFixed()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [10.98, 40.98],
          [36.96, 55.98],
          [21.96, 81.96],
          [-4.02, 66.96]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(30, 30),
        new DPoint(60, 30),
        new DPoint(60, 60),
        new DPoint(30, 60)
      ]);

      expect(p
        .loop()
        .rotate(Math.PI / 8)
        .toFixed()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [16.24, 39.2],
          [43.95, 50.68],
          [32.47, 78.39],
          [4.76, 66.91]
        ]);
    });
  });
  describe('round', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0.3),
        new DPoint(0.7, 1)
      ]);
      expect(p.loop()
        .round()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 0],
          [1, 1]
        ]);
    });
  });
  describe('flipVertically', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0.3),
        new DPoint(0.7, 1)
      ]);
      expect(p.loop()
        .flipVertically(10)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 9.7],
          [0.7, 9]
        ]);
    });
    test('2', () => {
      expect(DPolygon.parseFromWKT('POLYGON ((30.1 10.2, 40.3 40.4, 20.5 40.6, 10.7 20.8, 30.1 10.2))')
        .loop()
        .flipVertically(100)
        .run()
        .toWKT()).toBe('POLYGON ((30.1 89.8, 40.3 59.6, 20.5 59.4, 10.7 79.2, 30.1 89.8))');
    });
    test('3', () => {
      expect(DPolygon.parseFromWKT('POLYGON ((30.1 10.2, 40.3 40.4, 20.5 40.6, 10.7 20.8, 30.1 10.2),' +
        ' (25.2 18.3, 30.4 31.5, 21.6 32.7, 18.8 22.9, 25.1 18.2))')
        .loop()
        .flipVertically(new DPoint(100, 100))
        .run()
        .toWKT()).toBe('POLYGON ((30.1 89.8, 40.3 59.6, 20.5 59.4, 10.7 79.2, 30.1 89.8),' +
        ' (25.2 81.7, 30.4 68.5, 21.6 67.3, 18.8 77.1, 25.1 81.8))');
    });
  });
  describe('ceil', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0.3),
        new DPoint(0.7, 1)
      ]);
      expect(p.loop()
        .ceil()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 1],
          [1, 1]
        ]);
    });
  });
  describe('floor', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 0.3),
        new DPoint(0.7, 1)
      ]);
      expect(p.loop()
        .floor()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 0],
          [0, 1]
        ]);
    });
  });
  describe('abs', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, -0),
        new DPoint(-1, 1)
      ]);
      expect(p.loop()
        .abs()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 0],
          [1, 1]
        ]);
    });
  });
  describe('toFixed', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0.123, 0.345),
        new DPoint(0.567, 0.789)
      ]);
      expect(p.loop()
        .toFixed(1)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0.1, 0.3],
          [0.6, 0.8]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(0.123, 0.345),
        new DPoint(0.567, 0.789)
      ]);
      expect(p.loop()
        .toFixed()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0.12, 0.34],
          [0.57, 0.79]
        ]);
    });
  });
  describe('divide', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p.loop()
        .scale(2)
        .move(1)
        .divide(2)
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [1.5, 2.5],
          [3.5, 4.5]
        ]);
    });
    test('2', () => {
      const p = new DPolygon([
        new DPoint(1, 2),
        new DPoint(3, 4)
      ]);
      expect(p.loop()
        .scale(2)
        .move(1)
        .divide(new DPoint(2, 1))
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [1.5, 5],
          [3.5, 9]
        ]);
    });
  });
  describe('degreeToRadians', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .degreeToRadians()
        .toFixed()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 1.57],
          [3.14, 6.28]
        ]);
    });
  });
  describe('radiansToDegrees', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 1.57),
        new DPoint(3.14, 6.28)
      ]);
      expect(p
        .loop()
        .radiansToDegrees()
        .round()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 90],
          [180, 360]
        ]);
    });
  });
  describe('radiansToMeters', () => {
    test('1', () => {
      const p = new DPolygon([new DPoint(Math.PI / 4, Math.PI / 4)]);
      expect(p
        .loop()
        .radiansToMeters()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([[5009377.085, 5621521.485409545]]);
    });
  });
  describe('metersToRadians', () => {
    test('1', () => {
      const p = new DPolygon([new DPoint(5009377.085, 5621521.485409545)]);
      expect(p
        .loop()
        .metersToRadians()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([[Math.PI / 4, Math.PI / 4]]);
    });
  });
  describe('getHipPoint', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .getHipPoint()
        .toFixed()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [90, 90],
          [402.49, 402.49]
        ]);
    });
  });
  describe('getXPoint', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .getXPoint()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 0],
          [180, 180]
        ]);
    });
  });
  describe('getYPoint', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .getYPoint()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [90, 90],
          [360, 360]
        ]);
    });
  });
  describe('getWPoint', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .getWPoint()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [0, 0],
          [180, 180]
        ]);
    });
  });
  describe('getHPoint', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(180, 360)
      ]);
      expect(p
        .loop()
        .getHPoint()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [90, 90],
          [360, 360]
        ]);
    });
  });
  describe('setIfLessThan', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(0, 90),
        new DPoint(0, 0),
        new DPoint(90, 90),
        new DPoint(90, 0)
      ]);
      expect(p
        .loop()
        .setIfLessThan(new DPoint(45, 50))
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [45, 90],
          [45, 50],
          [90, 90],
          [90, 50]
        ]);
    });
  });
  describe('minus', () => {
    test('1', () => {
      const p = new DPolygon([
        new DPoint(-90, 90),
        new DPoint(-90, -90),
        new DPoint(90, 90),
        new DPoint(90, -90)
      ]);
      expect(p
        .loop()
        .minus()
        .run()
        .toArrayOfCoords())
        .toStrictEqual([
          [90, -90],
          [90, 90],
          [-90, -90],
          [-90, 90]
        ]);
    });
  });
});
