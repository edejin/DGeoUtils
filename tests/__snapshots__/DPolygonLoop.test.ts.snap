// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DPolygonLoop rotate 1 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAGF0lEQVR4nO3be2yV9R3H8fdzruUgp9AjttByEVLt2IhcNhcclxgR3JiXMG8QaZQQJbHI2JIxypY1A7L6Dyq4ELOgs4uYzSluIk6Jl4kQbyBRm5MgBqSEVlJrD1LoaXvOZ3+c55S6bEYN8Dttv6/m98+TtP1+v588z9P++isYY4wxxhhjjDHGGGOMMcYYY4wxxhhjjDHGGGOMMcYYY4wxxhhjzAAhLkNEXJdhAEQU8RGd0SOcGrrIdTlGrEF9Po6OOYD4juuyBicxGvEFQmyuEW0jcqF0h7I0VTQgRrgucXARDQiFnr5BIC0ruV/Rh+4R3aFcMB2xM5wYuQYRcl3qwCemIzKBdEhUHtRCnpZAn1GiJVW18l6Yf/Yx1ppooSO2wHXJA5fwEG8g5G1YowhpHaRSgt6VpErTrq8TH084G8yx8jcRE1yXP/CIaoTCLSUinlItG74URt/1TGSBEit/K1Lx/PslQ3PZnxFx120MDCKGOIoQ1Y+rlBaliP/fQARKE9HvRy1T5JE7RSaQC6a9+BStifsQAdct9W9iPUKRdyeLQEaPU/2VYfRdxxmlhdNrxRs/OvsYayk9wpmiWa7b6p/EpYgzZD0xc7em864yBL52IPn1jjddl9+yThwZlwsl64lj5a8gxrlusX8Rf0co2LBYHlntZuY3DiO/snj6S2yhiut+Ls4U5YLpjPZwrPwBxBDXrRY+cTVCwY6oGPuJltDwrcPouzqIaVXFLxRsWJy7U3LvlxTtxcsQnuu2C5MIIg4gxNr1itGhTxh7TgLJryYqNG9OrThwRd8fk5N0hae6br/wiOUIhY+WiViH1rP2nIbRd70SmKOx1evEp5fkQskExPHROxClrsdQGMRwxAmEuPkpjeGoOoidt0AEyhDQluG3K1a/QnRG89swaVpK1yGirkfiltiIUPi1qwTSU9x8XsPouz5nuJZV/VKBF+adfYydGHmCD7430fVY3BBViC6vJyCmvKc5vHbBwui7klRpdMM1uUD2TxWBTBI0xvV4LjyxE6HAlrsVpEfvMcVJIE0VaOgp/w6Z/W+B9oMG4W/4QqFUTJS2aDlbnIQh0KJtfhhP3p6/NNv1aNwQYtVGxUmphVInYeydgbws4vQQMe6IQE+6HoszkeQEEe7SRlY5CSMTQD9427876n4n0GnQONdzcecnz6uKpLoIOwnk0bv8MJoqRKxDoDrXI3FM2smPnYRxchgaddwPZNE2gZpAQ11PxKkF7HAShkC/ut8PY+8M4WUFsiNGSaqchHFoIop2+tsmV74l0B6QbTS6ujtu+Id/d2xdKlAGdKXrWRQEF2HsmuuHcXKYKGsWaKvrORSMCx1Gdwh990M/kNX1Ap0ElbmeQ8G47Y8JtSYuXCCba/wwDk0U0U6BVrueQWERinwe029WX6R05PyG0TYCJVr9QG58VqCPQYN8q/2/vXRtNr/tXdZYop3zguctkJrN/nfaNTd/6UbX7RcgXc6tf32Lxkm9f4+YsutiNU46t2E0TkKhbkRPUEx+X6CXXXde2Eo+u46165tpLxZCXldIdzxUovbicxPIvBf9uDetEKgHNNl1y/2AwlQla3n43jQ9QSFU1DpU9SuL1BP89mFsv8kPo22ESLQKtMl1p/2MEizYsY3dM3vfLxX7R+rV2d/8wFw6gioP+l+lZrNAbaCE6w77KU1lScOHHB7f+3656rkSHR7/9QP5w6/9z2ycJMJdAtW47qr/q2j6GXV1bXxxkRAKno5oeX2xTg776jBaSlE85Qcy/18CNYLCrtsZIDSE77+zgScWd+dPHw47Ftcj1UXKev87kDsf88PYflP+0nzXXQxAKuenz+1k74zex9ilb1+sPTO8L4WxbxoKZBDpiKg8KNB215UPbOGuH7Ji00c0l/Webr+uYaSay1DWQ7Ne9+OqXy1QGlTpuuRBQB4TD93DgytPkY4IoXBqiBY+4Z90by4T8ZRA9a4rHWQUZ9brf+Kf12d6Tx8KsXSrQM0g+9c2N1TJLX/bw/uTxb5pIpARaKnrqkxx+7WMP5wE7WNQnj4sSAqByl1XYYwxxhhjjDHGGGOMMcaYQe8/5OdtJdfL/kQAAAAASUVORK5CYII="`;

exports[`DPolygonLoop rotate 2 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAABmJLR0QA/wD/AP+gvaeTAAAGVElEQVR4nO3afWxV9R3H8fc5996WPlKFkUItDEFYO9nEziX9gyDxRgxjmhiq08EChtQ1LGLGBLIBkRADVWbdJLiOWLOiRLqZrEMmSxcxJiNxKzUjlsnGEHmwYbB19JG2997P/rinvcW4yKbcXx++r+aX3Jz79P19P/ndnJ7zA2OMMcYYY4wxxhhjjDHGGGOMMcYYY4wxxhhjjDHGGGOMMcYYM8qJsOsSzHDiORLeUUQlItt1OeObyEa0o+Cvd0InPVnPIma6Lm18EqsR8k/MFkdv11AwCS9BR95hxDcQvusyxw/RjBDL92oOJ/SlspdFbaXozk6F05N1ibi/FfEF1+WObaIcofDFAjGhV3tZLoGaKdPKghplrH1a/P3mVDCxUIzu7EZE1HXpY5OoR4jtGzWZi+plggRDo50C1fiPaWq0XjRUiIFwKpzO3A8RaxG5rqcxNojJiF4v7omZp7SR7VeFMXzE8dVEVEuLfib/yc3iwpRUMH0ZvfROqEPc6npKo5vYgJB/YIl84jrFzP8ayPBxjiJtytisiRV7RFM0FUxy1bQgKhAR19MbXYSPOIUQSw5qKQeuKYzho48MNVCh8pIXxU8eE525w08C2hkIP42Y7nqqo0PyVFbh00UiFNNBlvzPgQwfxynRd/OrlVn5U3FsXiqYuJ+gM/f3iCjCcz3tkUu8jhDrqzWD04oR+kyBDI7L5KuWSs0ue1XUrxD9kVQ4XTlnifsbETe6nv7IImYgYv6ViJhyQdWs/1zC+PhopkzLCp9XaMNT4sPpqWAGwgN05exH3Oa6FSOD2IGQV79cmVzRBaZcl0AGRxuFeiryhCYve0G8uUgkvNTP2bdfvst1O9wSmYh/IET5Ea2g/rqGMXz0E9EvWabp1d9KBtJ4r0AxUKHrtrgjViAUfvdWgXSE8rQFIlDcR1/8IPjpWnxIoEbXLXFLHEGI1Xt0G++mNQyBGu8Nwjg5S/hxgRa7bok74qsIhdtzRU6X9rA67YHc80YQyLqdAp0EjeOryOLnCFHzuApoVxc5aQ3j5CzkxxE9WWLSJYHWuW6JO2IioouEJ+a+r8epSfvq+MEzweqoWyVQD2iS67a4k7wqq1DTInkk9D5z0xpGTxaadCkI5I4/ClTnuiVuifcQ4v7XFKUp7avjpZVBGC3zBw/d4bol7oi7EAqfnyIi/XqN+9MeyNffCQJZVSdQi+uWuCV+hRBbtmoa59VPJK1htMwPwmgvENndAq1y3RJ3xFREvzcQEkXntJUtaV8dj7wYBLJznUDtoHG8zUg8iZC/f5nCDOgcRWkNo70AZXeTvH4154RAO123xB0RRpxDiDsP6wH2p311/Pj7weo4tFigBGiO67a4k7yVqvDxW4SX0GHuTGsYCQ/NOREEct+vBTrkuiVuicMI8b3nVcJxJfDSGsjv7g7COFMsQjGB7nPdknS6+pqQKAEWAjDpn6y88Vk8lNaCXqgKHtQ+CvHQWeD1tBYwooi76cv4aPAuXbgnoofrMvWnr6VndZwpRuEBRF+GKGwT6EeuW+Je5pXFLDl4gYYKEQsN3UItac5WbSXqzr5+gWzaFnzbvocE6mNc34S6inxQlPktv6V6fYKLk4eCyfp3hh6pDau19PMNoz+Cpp0PvmXB2wLtc92FEUqzyOt4hodf6bhqg1vCU3lTjhoq0ED4swey76Hgc1tLhZcQaIHrmY9wygRVsOjNd6itFF05Q+HccD5LT+zwdfam/z+QBW8Hn1a1W6BWkO3HunYqo7DtJap2X6G1dCiYUF9ISxuy1BRN/j9xrWG0liIvgejIE/mXBar69BrMJ9BE/PijfPM3pz++y734LzmqWeupM/fTA6naHbxr1xqBOkD5rmc2yskDRbnlrwfYuD3O2ZuGgsnsyNCq2oj+/JVPDqMjD+VfDl4975hAu1zPZozRNPI6fsjyvZdoiqY2uAmVNmfpF99JnlENBrJrTfDsWwsHD81zPYMxShmgCm4/+gd2bBD/umEomIltyZOA0zPQvGPB0QdfFegt11WPE5rL1I9qWPtcN+99eSgYL+YnH7UViow+gR50Xek4o3zQGu554wP2P5Da5b5tk0BtyVVlHNFCZv+tkS1bYxSfEWib64oMAJoK2gwqdl2JMcYYY4wxxhhjjDHGGGOMMcYYM8x/AKohfr/rwX6rAAAAAElFTkSuQmCC"`;
